src/content/docs/references/docs/optionals_advanced.md

```diff
@@ -0,0 +1,355 @@
+}
+```
+
+### Assert an Optional cannot have a missing `Result` and the program should panic with force unwrap `!!`
+
+If an `Empty` Optional is unexpected or cannot be easily handled we can panic, use sparingly:
+```c3
+int! optional_value = foo_may_error();
+if (catch excuse = optional_value) 
+{
+    unreachable("Unexpected excuse %s", excuse);
+}
+// Now optional_value is a regular variable
+```
+
+The force unwrap operator `!!` will exit the program if a `Excuse` is returned from `foo_may_error()`.
+```c3
+int regular_value = foo_may_error()!!;
+```
+
+## No void! variables
+
+The `void!` type has no possible representation as a variable, and may
+only be a function return type. 
+
+Use `if (catch)` to handle an `Empty` Optional or read the `Excuse`
+```c3
+fn void! test() 
+{
+    return IoError.FILE_NOT_FOUND?;
+}
+
+if (catch excuse = test()) // Retrieve underlying `Excuse`
+{
+    io::printfn("Excuse found: %s", excuse);
+    return excuse?;
+}
+```
+Or use `if (try)` to handle `Result` in Optional
+```c3
+fn void! test() 
+{
+    return;
+}
+
+if (try test()) // Handle `Result` outcome
+{
+    io::printn("successful test()");
+}
+```
+Store the `Excuse` returned from a `void!` function without `if (catch foo = optional_value)`, use the `@catch` macro to convert the result to an `anyfault`:
+```c3
+fn void! test() 
+{
+    return IoError.FILE_NOT_FOUND?;
+}
+
+anyfault excuse = @catch(test());
+```
+
+
+## Default assign when Optional is `Empty` using `??`
+If an expression returns a `Empty` Optional default assign using `??` operator.
+
+```c3
+import std::io;
+
+int regular_value;
+int! optional_value = function_may_error();
+if (catch optional_value) // A `Empty` Optional found in optional_value
+{   
+    regular_value = -1; // Assign default `Result` when `Empty` Optional
+}
+
+if (try optional_value;) // A `Result` was found in optional_value
+{
+    regular_value = optional_value;
+}
+```
+
+The operator `??` allows you to assign a default `Result` or `Excuse` when an expression contains a missing `Result`:
+
+```c3
+// Set default `Result` to -1 when `foo_may_error()` returns an `Empty` Optional
+int regular_value = foo_may_error() ?? -1;
+```
+
+### Signpost the unique origin of the same `Excuse`
+
+Catch a missing `Result` and change the underlying `Excuse` to a situation specific `Excuse`, that allows us to disitinguish one situation from the other, even with the same starting `Excuse`.
+
+```c3
+import std::io;
+
+fault NoHomework
+{
+    DOG_ATE_MY_HOMEWORK,
+    MY_TEXTBOOK_CAUGHT_FIRE,
+    DISTRACTED_BY_CAT_PICTURES
+}
+
+fn int! test() 
+{
+    return IoError.FILE_NOT_FOUND?;
+}
+
+fn void! main(String[] args) 
+{
+    int! a = test(); // IoError.FILE_NOT_FOUND
+    int! b = test(); // IoError.FILE_NOT_FOUND
+
+    // We can tell these appart by default assigning our own unique `Excuse`
+    // Our unique `Excuse` is assigned only if a `Excuse` is found
+    int c = test() ?? NoHomework.DOG_ATE_MY_HOMEWORK?;
```

Comment:
This is incorrect, maybe you meant:
`int! c = test() ?? NoHomework.DOG_ATE_MY_HOMEWORK?;`

Reply:
And same with "d" below.